src/hotspot/share/opto/memnode.cpp

```diff
@@ -2770,9 +2770,461 @@ Node *StoreNode::Ideal(PhaseGVN *phase, bool can_reshape) {
     }
   }
 
+#ifdef VM_LITTLE_ENDIAN
+  if (MergeStores && UseUnalignedAccesses) {
+    if (phase->C->post_loop_opts_phase()) {
+      Node* progress = Ideal_merge_primitive_array_stores(phase);
+      if (progress != nullptr) { return progress; }
+    } else {
+      phase->C->record_for_post_loop_opts_igvn(this);
+    }
+  }
+#endif
+
   return nullptr;                  // No further progress
 }
 
+// Link together multiple stores (B/S/C/I) into a longer one.
+Node* StoreNode::Ideal_merge_primitive_array_stores(PhaseGVN* phase) {
+  int opc = Opcode();
+  if (opc != Op_StoreB && opc != Op_StoreC && opc != Op_StoreI) {
+    return nullptr;
+  }
+
+  // Only merge stores on arrays.
+  if (adr_type()->isa_aryptr() == nullptr) {
+    return nullptr;
+  }
+
+  // If we can merge with use, then we must process use first.
+  StoreNode* use = can_merge_primitive_array_store_with_use(phase, true);
+  if (use != nullptr) {
+    return nullptr;
```

Comment:
Do you want to assert that the use is in the igvn worklist?

Reply:
Hmm. I think that would not be a good assert.
Let's assume we have 4 stores that would merge. Then the last one of them does the merging, and replaces itself with the merged store. If `this` is the second store, then it could merge with its def (the first store). But since it has a use that could also be merged with, we delegate the merging down. But it is not done by the 3rd store, rather the 4th. So we cannot assert that the 3rd would be in the worklist. The 3rd may have been processed before, and determined that it does not want to idealize itself, and be removed from the worklist.

Maybe I can improve the comment:
```
// Merging is done by the last store in a chain. We have a use that could be merged with, so we
// are not the last store, and hence must wait for some (recursive) use to do the merge.
```